workers/unity/Packages/com.improbable.gdk.testutils/TestLogDispatcher.cs

```diff
@@ -0,0 +1,148 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Improbable.Gdk.Core;
+using Improbable.Worker.Core;
+using NUnit.Framework;
+using UnityEngine;
+
+namespace Improbable.Gdk.TestUtils
+{
+    /// <summary>
+    ///     A ILogDispatcher implementation designed to be used in testing. This replaces the LogAssert approach with
+    ///     a more specialised one.
+    ///
+    ///     The expected usage is to use EnterExpectingScope() with a using block. This methods returns a Disposable
+    ///     object which you can mark logs as expected. When the object is disposed - it will assert against any logs.
+    /// </summary>
+    public class TestLogDispatcher : ILogDispatcher
+    {
+        private static readonly string AttemptToAccessConnectionError =
+            $"Cannot access or set the Connection in the {nameof(TestLogDispatcher)}";
+
+        private ExpectingScope currentExpectingScope;
+
+        // The connection will never be valid - so any attempt to get or set will throw.
+        public Connection Connection
+        {
+            get => throw new InvalidOperationException(AttemptToAccessConnectionError);
+            set => throw new InvalidOperationException(AttemptToAccessConnectionError);
+        }
+
+        public void HandleLog(LogType type, LogEvent logEvent)
+        {
+            if (currentExpectingScope != null)
+            {
+                currentExpectingScope.CheckIncomingLog(type, logEvent);
+                return;
+            }
+
+            if (type == LogType.Error || type == LogType.Exception)
+            {
+                Assert.Fail($"Encountered error log outside of an expecting scope: [{type}] - {logEvent}");
+            }
+        }
+
+        /// <summary>
+        ///     Creates and returns an disposable ExpectingScope object. This is intended to be used with a using block.
+        /// </summary>
+        /// <returns>An ExpectingScope instance.</returns>
+        /// <exception cref="InvalidOperationException">
+        ///     Throws if you already have an un-disposed ExpectingScope from this logger.
+        /// </exception>
+        public ExpectingScope EnterExpectingScope()
+        {
+            if (currentExpectingScope != null)
+            {
+                throw new InvalidOperationException("Cannot ");
+            }
+
+            currentExpectingScope = new ExpectingScope(this);
+            return currentExpectingScope;
+        }
+
+        public void Dispose()
```

Comment:
does this currently work? because internally the expecting scope never gets unset and we rely on people using the `using` statement. There should be a `ExitExpectingScope` method

Reply:
Disposing the `ExpectingScope` nulls its own reference in the `LogDispatcher` - perhaps that's a questionable pattern

Reply:
Could add a helper method

Reply:
Realised I put a throw inside `Dispose` will change that around